### ISSUE TITLE
Behaviour of return type in function types is ambiguous

### ISSUE THREAD

**Issue by sushantbs** [Microsoft/TypeScript#28151]
I have gone through the documentation but couldn't find anything that explains this behavior. I apologize if this is a non issue.
  

**TypeScript Version:**  3.0.3

**Search Terms:** function types

**Code**

```ts
type Ticket = {
	testStr: string;
	testNum: number;
};

type TicketChanger = (
	param1: Ticket,
	param2: { name: "testStr" | "testNum"; value: string | number }
) => Ticket;

let ticket = {
	testStr: "test",
	testNum: 10
};

const badFn: TicketChanger = (ticket, { name, value }) => {
	return {
		...ticket,
		[ticket[name]]: value,
		extra: "This should be an error, but is allowed"
	};
};

const goodFn: TicketChanger = (ticket, { name, value }): Ticket => {
	
	return {
		...ticket,
		[ticket[name]]: value,
		extra: "This should be an error, and actually is"
	};
};

let newStrTicket: Ticket = badFn(ticket, { name: "testStr", value: "newTest" });
let newNumTicker: Ticket = goodFn(ticket, { name: "testNum", value: 100 });

let inferredTicket = badFn(ticket, { name: "testNum", value: "newTest" });

```

**Expected behavior:**
The `badFn` function should show a ts error on the return object as the function has been assigned the `TicketChanger` type 

**Actual behavior:**
No ts error in `badFn`.

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=type%20Ticket%20%3D%20%7B%0D%0A%09testStr%3A%20string%3B%0D%0A%09testNum%3A%20number%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20TicketChanger%20%3D%20(%0D%0A%09param1%3A%20Ticket%2C%0D%0A%09param2%3A%20%7B%20name%3A%20%22testStr%22%20%7C%20%22testNum%22%3B%20value%3A%20string%20%7C%20number%20%7D%0D%0A)%20%3D%3E%20Ticket%3B%0D%0A%0D%0Alet%20ticket%20%3D%20%7B%0D%0A%09testStr%3A%20%22test%22%2C%0D%0A%09testNum%3A%2010%0D%0A%7D%3B%0D%0A%0D%0Aconst%20badFn%3A%20TicketChanger%20%3D%20(ticket%2C%20%7B%20name%2C%20value%20%7D)%20%3D%3E%20%7B%0D%0A%09return%20%7B%0D%0A%09%09...ticket%2C%0D%0A%09%09%5Bticket%5Bname%5D%5D%3A%20value%2C%0D%0A%09%09extra%3A%20%22This%20should%20be%20an%20error%2C%20but%20is%20allowed%22%0D%0A%09%7D%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20goodFn%3A%20TicketChanger%20%3D%20(ticket%2C%20%7B%20name%2C%20value%20%7D)%3A%20Ticket%20%3D%3E%20%7B%0D%0A%09%0D%0A%09return%20%7B%0D%0A%09%09...ticket%2C%0D%0A%09%09%5Bticket%5Bname%5D%5D%3A%20value%2C%0D%0A%09%09extra%3A%20%22This%20should%20be%20an%20error%2C%20and%20actually%20is%22%0D%0A%09%7D%3B%0D%0A%7D%3B%0D%0A%0D%0Alet%20newStrTicket%3A%20Ticket%20%3D%20badFn(ticket%2C%20%7B%20name%3A%20%22testStr%22%2C%20value%3A%20%22newTest%22%20%7D)%3B%0D%0Alet%20newNumTicker%3A%20Ticket%20%3D%20goodFn(ticket%2C%20%7B%20name%3A%20%22testNum%22%2C%20value%3A%20100%20%7D)%3B%0D%0A%0D%0Alet%20inferredTicket%20%3D%20badFn(ticket%2C%20%7B%20name%3A%20%22testNum%22%2C%20value%3A%20%22newTest%22%20%7D)%3B%0D%0A

**Related Issues:**

**Comment by fatcerberus** [comment]
Just to drive the point home, you can make `goodfn` not produce an error either by rewriting it as:
```ts
const goodFn: TicketChanger = (ticket, { name, value }): Ticket => {
	let o = {
		...ticket,
		[ticket[name]]: value,
		extra: "This should be an error, and actually is"
	};
	return o;
};
```

**Comment by DanielRosenwasser** [comment]
Correct; the problem you're observing is that we still need to infer `badFn`'s type, and to do that we widen the return expressions. That return expression starts out as a fresh object literal, and then loses its freshness after widening. We only perform excess property checking on fresh object literal types. While it's a bit out of date, you can read more here: https://github.com/Microsoft/TypeScript/blob/0e372e98e83e649a1a7c103ea3dd4df8ce5f51bb/doc/spec.md#3115-excess-properties

**Comment by sushantbs** [comment]
Thanks a lot for clarifying this guys. I'm probably need to read about this in more detail cuz some of the things still aren't very clear. Thanks for pointing me in the right direction @DanielRosenwasser.

**Comment by fatcerberus** [comment]
@sushantbs If it makes it any clearer why you don't get an error on `badFn`, it's the same case as:
```js
const badFn = (ticket, { name, value }) => {
	return {
		...ticket,
		[ticket[name]]: value,
		extra: "random extra stuff that doesn't belong"
	};
};
const fn: TicketChanger = badFn;  // look ma, no errors!
```

The definition of the function itself is logically a separate thing from the assignment.
